File: 1022. Sum of Root To Leaf Binary Numbers/Solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def sumRootToLeaf_MK1(self, root: TreeNode) -> int:
        '''
        Recursive Preorder Traversal.
        Time complexity: O(N). N is a number of nodes.
        Space complexity: O(H). H is a tree height.
        '''
        def dfs(root: TreeNode, curNum: int) -> int:
            if not root:
                return 0
            curNum = (curNum << 1) + root.val
            if not root.left and not root.right:
                return curNum
            return dfs(root.left, curNum) + dfs(root.right, curNum)
        return dfs(root, 0)

    def sumRootToLeaf_MK2(self, root: TreeNode) -> int:
        '''
        Iterative Preorder Traversal.
        Time complexity: O(N). N is a number of nodes.
        Space complexity: O(H). H is a tree height.
        '''
        if not root:
            return 0
        ret = 0
        stack = [(root, 0)]
        while stack:
            node, curNum = stack.pop()
            curNum = (curNum << 1) + node.val
            if not node.left and not node.right:
                ret += curNum
            else:
                if node.right:
                    stack.append((node.right, curNum))
                if node.left:
                    stack.append((node.left, curNum))
        return ret

    def sumRootToLeaf_MK3(self, root: TreeNode) -> int:
        '''
        Morris Preorder Traversal.
        Time complexity: O(N). N is a number of nodes.
        Space complexity: O(1).
        '''
        ret, curNum, steps = 0, 0, 0
        while root:
            if not root.left:
                curNum = (curNum << 1) + root.val
                if not root.right:
                    ret += curNum
                root = root.right
            else:
                succ = root.left
                steps = 1
                while succ.right and succ.right != root:
                    succ = succ.right
                    steps += 1
                if succ.right == root:
                    succ.right = None
                    if not succ.left:
                        ret += curNum
                    for _ in range(steps):
                        curNum >>= 1
                    root = root.right
                else:
                    succ.right = root
                    curNum = (curNum << 1) + root.val
                    root = root.left
        return ret

File: 1022. Sum of Root To Leaf Binary Numbers/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    /**
     * Recursive Preorder Traversal.
     * Time complexity: O(N). N is a number of nodes.
     * Space complexity: O(H). H is a tree height.
     */
    public int sumRootToLeaf_MK1(TreeNode root) {
        return this.dfs(root, 0);
    }

    /**
     * Morris Preorder Traversal.
     * Time complexity: O(N). N is a number of nodes.
     * Space complexity: O(1).
     */
    public int sumRootToLeaf_MK2(TreeNode root) {
        int ret = 0, curNum = 0, steps = 0;
        while (root != null) {
            if (root.left == null) {
                curNum = (curNum << 1) + root.val;
                root = root.right;
                if (root == null)
                    ret += curNum;
            } else {
                TreeNode succ = root.left;
                steps = 1;
                while (succ.right != null && succ.right != root) {
                    succ = succ.right;
                    steps += 1;
                }
                if (succ.right == null) {
                    succ.right = root;
                    curNum = (curNum << 1) + root.val;
                    root = root.left;
                } else {
                    succ.right = null;
                    if (succ.left == null)
                        ret += curNum;
                    for (int i = 0; i < steps; i++)
                        curNum >>= 1;
                    root = root.right;
                }
            }
        }
        return ret;
    }

    private int dfs(TreeNode root, int curNum) {
        if (root == null)
            return 0;
        curNum = (curNum << 1) + root.val;
        return root.left == null && root.right == null ? curNum : this.dfs(root.left, curNum) + this.dfs(root.right, curNum);
    }
}
